src/crypto/tls/ech.go

```diff
@@ -0,0 +1,821 @@
+// 32-byte string, which will be interpreted as an X25519 public key.
+var echDummyPublicKey = []byte{
+	143, 38, 37, 36, 12, 6, 229, 30, 140, 27, 167, 73, 26, 100, 203, 107, 216,
+	81, 163, 222, 52, 211, 54, 210, 46, 37, 78, 216, 157, 97, 241, 244,
+}
+
+func (c *Conn) echOfferOrBypass(helloBase *clientHelloMsg) (hello, helloInner *clientHelloMsg, err error) {
+	config := c.config
+
+	// Trigger failure scenarios, if testing.
+	echConfigs := config.ClientECHConfigs
+	if c.hrrTriggered && testingTriggerECHBypassAfterHRR {
+		echConfigs = nil
+	}
+	if !c.hrrTriggered && testingTriggerECHBypassBeforeHRR {
+		echConfigs = nil
+	}
+
+	// Select an ECH configuration suitable for this connection.
+	var echConfig *ECHConfig
+	for _, conf := range echConfigs {
+		// A suitable configuration is one that offers an HPKE ciphersuite that
+		// is supported by the client and indicates the version of ECH
+		// implemented by this TLS client.
+		if conf.isSupported() && extensionECH == conf.version {
+			echConfig = &conf
+			break
+		}
+	}
+
+	hello = new(clientHelloMsg)
+	*hello = *helloBase
+
+	// Decide whether to offer the ECH extension in this connection, If offered,
+	// then hello is set to the ClientHelloOuter and helloInner is set to the
+	// ClientHelloInner.
+	if echConfig != nil && c.config.maxSupportedVersion() >= VersionTLS13 {
+		helloInner = new(clientHelloMsg)
+		*helloInner = *helloBase
+
+		// Prepare the ClientHelloOuter.
+		//
+		// Offer an empty "encrypted_client_hello" extension.
+		helloInner.encryptedClientHelloOffered = true
+
+		// Only offer to negotiate TLS 1.3 and above.
+		//
+		// TODO(cjpatton): In the spec, should this be a SHOULD instead of a
+		// MUST?
+		helloInner.supportedVersions = config.supportedVersionsFromMin(VersionTLS13)
+
+		// Prepare the ClientHelloOuter.
+		//
+		// Set the "server_name" to be the client-facing server.
+		hello.serverName = hostnameInSNI(string(echConfig.rawPublicName))
+
+		// Generate a fresh ClientHelloOuter.random.
+		hello.random = make([]byte, 32)
+		_, err = io.ReadFull(config.rand(), hello.random)
+		if err != nil {
+			return nil, nil, fmt.Errorf("tls: %s", err)
+		}
+
+		// Prepare the encryption context. Note that c.ech.hrrPsk is initially
+		// nil, meaning no PSK is used for encrypting the first ClientHelloInner
+		// in case of HRR.
+		ctx, enc, err := echConfig.setupClientContext(c.ech.hrrPsk, config.rand())
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// Compress the ClientHelloInner using the "outer_extension" mechanism.
+		//
+		// NOTE(cjpatton): It would be nice to incorporate more extensions, but
+		// "key_share" is the last extension to appear in the ClientHello before
+		// "pre_shared_key". As a result, the only contiguous sequence of outer
+		// extensions that contains "key_share" is "key_share" itself. Note that
+		// we cannot change the order of extensions in the ClientHello, as the
+		// unit tests appear to be ossified on the extension order.
```

Comment:
"ossified" is a strong word with negative connotations, the extension order is only fixed due to an implementation detail of the client/server, and the unit tests themselves that replay protocol messages. If both of them change, the protocol operates as usual.

Can we use a different word here?

Reply:
Changed to "as the unit tests expect 'key_share' to be the second to last extension".